packages/mongodb-dialects/spring-criteria/src/test/kotlin/com/mongodb/jbplugin/dialects/springcriteria/IntegrationTest.kt

```diff
@@ -0,0 +1,210 @@
+        }
+    }
+
+    override fun interceptTestMethod(
+        invocation: InvocationInterceptor.Invocation<Void>,
+        invocationContext: ReflectiveInvocationContext<Method>,
+        extensionContext: ExtensionContext,
+    ) {
+        val throwable: AtomicReference<Throwable?> = AtomicReference(null)
+        val finished = AtomicBoolean(false)
+
+        val fixture = extensionContext.getStore(namespace).get(testFixtureKey) as CodeInsightTestFixture
+        val dumbService = DumbService.getInstance(fixture.project)
+
+        dumbService.runWhenSmart {
+            val result =
+                runCatching {
+                    invocation.proceed()
+                }
+            result.onSuccess {
+                finished.set(true)
+            }
+
+            result.onFailure {
+                finished.set(true)
+                throwable.set(it)
+            }
+        }
+
+        while (!finished.get()) {
+            Thread.sleep(1)
+        }
+
+        throwable.get()?.let {
+            it.printStackTrace(System.err)
+            throw it
+        }
+    }
+
+    override fun afterAll(context: ExtensionContext) {
+        val testFixture = context.getStore(namespace).get(testFixtureKey) as CodeInsightTestFixture
+
+        ApplicationManager.getApplication().invokeAndWait {
+            testFixture.tearDown()
+        }
+    }
+
+    override fun supportsParameter(
+        parameterContext: ParameterContext,
+        extensionContext: ExtensionContext,
+    ): Boolean =
+        parameterContext.parameter.type == Project::class.java ||
+            parameterContext.parameter.type == CodeInsightTestFixture::class.java ||
+            parameterContext.parameter.type == PsiFile::class.java ||
+            parameterContext.parameter.type == JavaPsiFacade::class.java
+
+    override fun resolveParameter(
+        parameterContext: ParameterContext,
+        extensionContext: ExtensionContext,
+    ): Any {
+        val fixture = extensionContext.getStore(namespace).get(testFixtureKey) as CodeInsightTestFixture
+
+        return when (parameterContext.parameter.type) {
+            Project::class.java -> fixture.project
+            CodeInsightTestFixture::class.java -> fixture
+            PsiFile::class.java -> fixture.file
+            JavaPsiFacade::class.java -> JavaPsiFacade.getInstance(fixture.project)
+            else -> TODO("Parameter of type ${parameterContext.parameter.type.canonicalName} is not supported.")
+        }
+    }
+
+    private fun pathToClassJarFile(javaClass: Class<*>): String {
+        val classResource: URL =
+            javaClass.getResource(javaClass.getSimpleName() + ".class")
+                ?: throw RuntimeException("class resource is null")
+        val url: String = classResource.toString()
+        if (url.startsWith("jar:file:")) {
+            // extract 'file:......jarName.jar' part from the url string
+            val path = url.replace("^jar:(file:.*[.]jar)!/.*".toRegex(), "$1")
```

Comment:
The `!` is a literal `!` here and not some special syntax, correct?

Are we okay with this regexp potentially mis-matching if `.jar` occurs in the rest of the URL again?

Reply:
You are right that `!` is a literal `!`, but in theory the Regex is bound to the beginning (hence the ^). However it's forbidden to have a Java class with a dot in the name so it shouldn't happen unless someone adds a malicious JAR to our testing environment. If you think I should fix it I'm fine, I think it can be easy to fix it.

Reply:
All good, that makes sense then 👍